pkg/ipamd/datastore/data_store.go

```diff
@@ -974,21 +974,28 @@ func (e *ENI) hasPods() bool {
 }
 
 // GetENINeedsIP finds an ENI in the datastore that needs more IP addresses allocated
-func (ds *DataStore) GetENINeedsIP(maxIPperENI int, skipPrimary bool) *ENI {
+func (ds *DataStore) GetENINeedsIP(maxIPperENI int, skipPrimary bool, useSubnetDiscovery bool, insufficientENI map[*ENI]bool) (*ENI, int) {
 	ds.lock.Lock()
 	defer ds.lock.Unlock()
 	for _, eni := range ds.eniPool {
 		if (skipPrimary && eni.IsPrimary) || eni.IsTrunk {
 			ds.log.Debugf("Skip needs IP check for trunk ENI of primary ENI when Custom Networking is enabled")
 			continue
 		}
+		if _, ok := insufficientENI[eni]; ok {
+			continue
+		}
 		if len(eni.AvailableIPv4Cidrs) < maxIPperENI {
 			ds.log.Debugf("Found ENI %s that has less than the maximum number of IP/Prefixes addresses allocated: cur=%d, max=%d",
 				eni.ID, len(eni.AvailableIPv4Cidrs), maxIPperENI)
-			return eni
+			if useSubnetDiscovery {
+				return eni, len(ds.eniPool)
+			}
+			return eni, 1
 		}
 	}
-	return nil
+	// Returning 1 since we want one iteration of IP allcocation if subnet discovery is off
+	return nil, 1
```

Comment:
Can this scenario be covered using unit-test (return 1 iteration of ip allocation). Same for line 994.